docs/datatypes/stm.md

```diff
@@ -0,0 +1,115 @@
+more locks are introduced into the equation.
+
+Transactional memory eliminates many of the difficulties that affect low-level lock-based programming since transactional
+data structures are lock-free.
+
+## Transactional data structures
+
+There are a variety of transactional data structures that can take part in an STM transaction. These include:
+* [TRef](tref.md): A mutable reference to an immutable value
+* [TPromise](tpromise.md): A mutable reference that can be set exactly once 
+* [TArray](tarray.md): An array of mutable references
+* [TQueue](tqueue.md): A mutable queue
+* [TSet](tset.md): A mutable set
+* [TSemaphore](tsemaphore.md): A semaphore 
+* [TReentrantLock](treentrantlock.md): A reentrant lock
+
+## The STM datatype
+
+An `STM[E, A]` represents an effect that can be performed transactionally resulting in a failure `E` or a success `A`.
+The `STM` data-type is _not_ as powerful as the `ZIO[R, E, A]` datatype as it does not allow you to perform arbitrary 
+effects. These are because actions inside STM actions can be executed an arbitrary amount of times (and rolled-back as 
+well). Only STM actions and pure computation may be performed inside a memory transaction. No STM actions can be 
+performed outside a transaction, so you cannot accidentally read or write a transactional data structure outside the
+protection of `STM.atomically` (or without explicitly `commit`ting the transaction). For example:
+
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+def transferMoney(from: TRef[Long], to: TRef[Long], amount: Long): STM[String, Long] =
+  for {
+    senderBal <- from.get
+    _         <- if (senderBal < amount) STM.fail("Not enough money")
+                 else STM.unit
+    _         <- from.update(existing => existing - amount)
+    _         <- to.update(existing => existing + amount)
+    recvBal   <- to.get
+  } yield recvBal
+
+val program: IO[String, Long] = for {
+  sndAcc  <- STM.atomically(TRef.make(1000L))
+  rcvAcc  <- STM.atomically(TRef.make(0L))
+  recvAmt <- STM.atomically(transferMoney(sndAcc, rcvAcc, 500L))
+} yield recvAmt
+```
+
+`transferMoney` describes an atomic transfer process between a sender and a receiver. The transaction will fail if the
+sender does not have enough of money in their account. This means that individual accounts will be debited and credited
+atomically. If the transaction fails in the middle, the entire process will be rolled back, and it will appear that 
+nothing has happened. Here, we see that `STM` effects compose using a for-comprehension and that wrapping an `STM` effect
+with `STM.atomically` (or calling `commit` on any STM effect) turns the `STM` effect into a `ZIO` effect which can be
+executed. STM transactions compose sequentially. By using `STM.atomically` (or `commit`), the programmer identifies 
+atomic transaction in the sense that the entire set of operations within `STM.atomically` appears to take place 
+indivisibly.
+
+## Exceptions
```

Comment:
They are not necessary exceptions. Maybe state errors as in error channel?

Reply:
reworded